servers/src/main/java/tachyon/worker/block/BlockMasterSync.java

```diff
@@ -186,13 +191,8 @@ private void handleMasterCommand(Command cmd) throws Exception {
       // Master requests blocks to be removed from Tachyon managed space.
       case Free:
         for (long block : cmd.mData) {
-          try {
-            mBlockDataManager.removeBlock(Users.MASTER_COMMAND_USER_ID, block);
-          } catch (NotFoundException nfe) {
-            // TODO: Throw a better exception here
-            // NotFoundException will be thrown if the block is unable to be locked.
-            LOG.warn("Failed master free block cmd for: " + block + " due to concurrent read.");
-          }
+          mFixedExecutionService.execute(new BlockRemover(mBlockDataManager,
```

Comment:
Thank you for raising up this question. The motivation of this change are
two folds: (1) changing the removing block operation to async ( fire and go
); (2) parallelization the removing operation.

Given that, we picked fixed thread pool, which has bounded resource
consumption (currently 10, which can be adjusted by size) and able to
parallel the operation.  Employing just one remove thread will able to
reduce the resource cost for sure, but we will lose the parallelization.

Usually, the jvm performance may get noticeable impacted or even run into
problem when starting hundreds of thread or unbounded number of threads.
But they may don't have direct link due to the creation of thread
themselves, most likely due the operation in the thread.

In this change, we ran through integration test to check the removal
correctness. If there's available performance measurement report of the
master ( or  anyone noticed ) shows that the performance/cost of the master
is impacted due to the removal threads, we do need to adjust the pool size.

Thanks!
Zhen



On Mon, Aug 10, 2015 at 12:42 AM, Shiqing Fan <notifications@github.com>
wrote:

> In servers/src/main/java/tachyon/worker/block/BlockMasterSync.java
> <https://github.com/amplab/tachyon/pull/1252#discussion_r36607858>:
>
> > @@ -186,13 +191,8 @@ private void handleMasterCommand(Command cmd) throws Exception {
> >        // Master requests blocks to be removed from Tachyon managed space.
> >        case Free:
> >          for (long block : cmd.mData) {
> > -          try {
> > -            mBlockDataManager.removeBlock(Users.MASTER_COMMAND_USER_ID, block);
> > -          } catch (NotFoundException nfe) {
> > -            // TODO: Throw a better exception here
> > -            // NotFoundException will be thrown if the block is unable to be locked.
> > -            LOG.warn("Failed master free block cmd for: " + block + " due to concurrent read.");
> > -          }
> > +          mFixedExecutionService.execute(new BlockRemover(mBlockDataManager,
>
> @zhljen <https://github.com/zhljen> @calvinjia
> <https://github.com/calvinjia> One more question, is that really good
> enough to create a single thread for each block(here is up to
> DEFAULT_BLOCK_REMOVER_POOL_SIZE concurrent block remover thread) remove
> operation in the for-loop? It may bring considerable resource consumption.
> How about just pass the cmd.mData list to the one block-remove-thread ?
> The remove-thread will also execute asynchronous without blocking the
> handleMasterCommand method but with less resources cost. Thanks!
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/amplab/tachyon/pull/1252/files#r36607858>.
>